src/Cli/dotnet/Commands/Test/Terminal/TerminalTestReporter.cs

```diff
@@ -152,28 +154,52 @@ public TerminalTestReporter(IConsole console, TerminalTestReporterOptions option
         _terminalWithProgress = terminalWithProgress;
     }
 
-    public void TestExecutionStarted(DateTimeOffset testStartTime, int workerCount, bool isDiscovery, bool isHelp)
+    public void TestExecutionStarted(DateTimeOffset testStartTime, int workerCount, bool isDiscovery, bool isHelp, bool isRetry)
     {
         _isDiscovery = isDiscovery;
         _isHelp = isHelp;
+        _isRetry = isRetry;
         _testExecutionStartTime = testStartTime;
         _terminalWithProgress.StartShowingProgress(workerCount);
     }
 
-    public void AssemblyRunStarted(string assembly, string? targetFramework, string? architecture, string? executionId)
+    public void AssemblyRunStarted(string assembly, string? targetFramework, string? architecture, string? executionId, string? instanceId)
     {
+        var assemblyRun = GetOrAddAssemblyRun(assembly, targetFramework, architecture, executionId);
+        assemblyRun.Tries.Add(instanceId);
+
```

Comment:
`_isRetry |= assemblyRun.Tries.Count > 1`? (so that if we failed to detect the `--retry-failed-tests` we are still able to have reasonable experience?

Reply:
I didn't realize auto-merge was enabled. @nohwnd Let me know what you think of this suggestion and probably follow-up to do the change.

Reply:
yeah sounds great.

Reply:
https://github.com/dotnet/sdk/pull/48142